src/main/java/com/openlattice/datastore/data/controllers/DataController.java

```diff
@@ -189,12 +189,55 @@
         if ( authz.checkIfHasPermissions( new AclKey( entitySetId ),
                 Principals.getCurrentPrincipals(),
                 EnumSet.of( Permission.READ ) ) ) {
-            return loadNormalEntitySetData( entitySetId, selection );
+            EntitySet es = edmService.getEntitySet( entitySetId );
+            if(es.isLinking()) {
+                return loadLinkedEntitySetData( entitySetId, selection, es );
+            } else {
+                return loadNormalEntitySetData( entitySetId, selection );
+            }
         } else {
             throw new ForbiddenException( "Insufficient permissions to read the entity set or it doesn't exists." );
         }
     }
 
+    private EntitySetData<FullQualifiedName> loadLinkedEntitySetData(
+            UUID linkingEntitySetId,
+            EntitySetSelection selection,
+            EntitySet es) {
+        Set<UUID> allEntitySetIds = Sets.newHashSet( es.getLinkedEntitySets());
+        checkState( !allEntitySetIds.isEmpty(),
+                "Linked entity sets are empty for linking entity set %s", linkingEntitySetId );
+
+        EntityType linkedEntityType = edmService.getEntityTypeByEntitySetId( allEntitySetIds.iterator().next() );
+        checkState( allEntitySetIds.stream().skip( 1 )
```

Comment:
We shouldn't check this for every read. We should ensure in the linking controller that you can't add an entity set of the wrong type. This adds overhead to every single read call.